### ISSUE TITLE
flaky test: osquery_logger_tests-test segfaults

### ISSUE THREAD

**Issue by theopolis** [osquery/osquery#6928]
For reference, while the job results still exist: https://dev.azure.com/trailofbits/osquery/_build/results?buildId=4966&view=logs&jobId=414a5795-8c4a-54f4-80f7-b179ccc11220&j=414a5795-8c4a-54f4-80f7-b179ccc11220&t=a95a6d7c-f64e-50a1-0728-2bc9d3abd5ac

```
11: Test command: /Users/runner/work/1/b/build/osquery/logger/tests/osquery_logger_tests-test
11: Test timeout computed to be: 10000000
11: Running main() from /Users/runner/work/1/b/build/libs/src/patched-source/googletest/src/googletest/src/gtest_main.cc
11: [==========] Running 12 tests from 1 test case.
11: [----------] Global test environment set-up.
11: [----------] 12 tests from LoggerTests
11: [ RUN      ] LoggerTests.test_plugin
11: [       OK ] LoggerTests.test_plugin (4 ms)
11: [ RUN      ] LoggerTests.test_logger_init
11: W0128 19:09:26.578464 238435776 logger.cpp:150] Logger test is generating a warning status (1)
11: [       OK ] LoggerTests.test_logger_init (1 ms)
11: [ RUN      ] LoggerTests.test_log_string
11: [       OK ] LoggerTests.test_log_string (0 ms)
11: [ RUN      ] LoggerTests.test_logger_log_status
11: W0128 19:09:26.580158 238435776 logger.cpp:184] Logger test is generating a warning status (2)
11: [       OK ] LoggerTests.test_logger_log_status (0 ms)
11: [ RUN      ] LoggerTests.test_logger_status_level
11: I0128 19:09:26.580386 238435776 logger.cpp:200] Logger test is generating an info status
11: W0128 19:09:26.580579 238435776 logger.cpp:208] Logger test is generating a warning status
11: W0128 19:09:26.580754 238435776 logger.cpp:217] Logger test is generating a warning status
11: W0128 19:09:26.581029 238435776 logger.cpp:226] Logger test is generating a warning status
11: [       OK ] LoggerTests.test_logger_status_level (1 ms)
11: [ RUN      ] LoggerTests.test_feature_request
11: [       OK ] LoggerTests.test_feature_request (0 ms)
11: [ RUN      ] LoggerTests.test_logger_variations
11: W0128 19:09:26.581575 238435776 logger.cpp:258] Logger test is generating a warning status (3)
11: [       OK ] LoggerTests.test_logger_variations (0 ms)
11: [ RUN      ] LoggerTests.test_logger_snapshots
11: [       OK ] LoggerTests.test_logger_snapshots (1 ms)
11: [ RUN      ] LoggerTests.test_multiple_loggers
11: W0128 19:09:26.582305 238435776 logger.cpp:328] Logger test is generating a warning status (4)
11: W0128 19:09:26.582564 238435776 logger.cpp:336] Logger test is generating a warning status (5)
11: [       OK ] LoggerTests.test_multiple_loggers (0 ms)
11: [ RUN      ] LoggerTests.test_logger_scheduled_query
11: [       OK ] LoggerTests.test_logger_scheduled_query (1 ms)
11: [ RUN      ] LoggerTests.test_logger_numeric_flag
11: [       OK ] LoggerTests.test_logger_numeric_flag (0 ms)
11: [ RUN      ] LoggerTests.test_recursion
11: W0128 19:09:26.583528 238435776 logger.cpp:461] Log to the recursive logger
11: W0128 19:09:26.583667 238435776 logger.cpp:465] recurse
11: W0128 19:09:26.583958 177364992 logger.cpp:433] Logging a status within a status logger
11/82 Test #11: osquery_logger_tests-test .............................................***Exception: SegFault  0.66 sec

```

**Comment by Smjert** [comment]
Not sure what's happening here, I feared it was my recent rewrite of the glog CMake, but I re-checked the generated `config.h` against the one that the original CMake was creating, and it matches.

I've also run for a while the test built in Release mode from the latest master on my VM with 10.15, then also in Debug mode, then Release and Debug mode plus the thread sanitizer, then removed the thread sanitizer and compiled with the address sanitizer, but I couldn't hit it.

**Comment by theopolis** [comment]
I did the same test on my local VMs, which have been rather good at crashing, but I was unable to reproduce.

**Comment by Smjert** [comment]
I've reopened because it's still happening also on Github Actions.
I've been able to reproduce and find the issue, although the fix for the crash then triggers another issue which I'm trying to investigate.

Slowing down processing helps a bit in reproducing the crash, so setting 2 cores on a VM helps; also running the test via ctest somehow helps too.
In any case the first issue is here (in the BufferedLogSink::WaitTillSent implementation): https://github.com/osquery/osquery/blob/85a8e1fcb5c69bd5930630cf50d640e9e9c986b7/osquery/logger/logger.cpp#L346-L347
`senders` is a queue, and we are getting the **last** element (which is a `std::future`), to later use it, and then we are removing the **front** element.
Later the future is waited on; now `senders` can have multiple elements inside, so getting out the last element but removing the front one, means that the last element potentially is waited on multiple times.
When a future gets "used", it becomes invalid so using (waiting) on it more than once is UB.

Now, substituing `back()` with `front()` removes the crash but now the test fails various assertions (not always, but I believe it's when there are more than once future queued):
```
11: [ RUN      ] LoggerTests.test_recursion
11: W0429 14:46:39.431454 310427072 logger.cpp:464] Log to the recursive logger
11: W0429 14:46:39.431514 310427072 logger.cpp:468] recurse
11: W0429 14:46:39.431773 135708672 logger.cpp:436] Logging a status within a status logger
11: /Users/smjert/Development/osquery/src/osquery/logger/tests/logger.cpp:477: Failure
11: Expected equality of these values:
11:   3U
11:     Which is: 3
11:   plugin->statuses
11:     Which is: 2
11: W0429 14:46:39.432055 310427072 logger.cpp:482] recurse
11: /Users/smjert/Development/osquery/src/osquery/logger/tests/logger.cpp:485: Failure
11: Expected equality of these values:
11:   3U
11:     Which is: 3
11:   plugin->statuses
11:     Which is: 2
11: W0429 14:46:39.432219 310427072 logger.cpp:436] Logging a status within a status logger
11: /Users/smjert/Development/osquery/src/osquery/logger/tests/logger.cpp:489: Failure
11: Expected equality of these values:
11:   4U
11:     Which is: 4
11:   plugin->statuses
11:     Which is: 3
11: /Users/smjert/Development/osquery/src/osquery/logger/tests/logger.cpp:491: Failure
11: Expected equality of these values:
11:   5U
11:     Which is: 5
11:   plugin->statuses
11:     Which is: 4
11: [  FAILED  ] LoggerTests.test_recursion (1 ms)
11: [----------] 12 tests from LoggerTests (4 ms total)
11:
11: [----------] Global test environment tear-down
11: [==========] 12 tests from 1 test case ran. (4 ms total)
11: [  PASSED  ] 11 tests.
11: [  FAILED  ] 1 test, listed below:
11: [  FAILED  ] LoggerTests.test_recursion
11:
11:  1 FAILED TEST
1/1 Test #11: osquery_logger_tests-test ........Subprocess aborted***Exception:  14.00 sec
```

**Comment by theopolis** [comment]
> was the interest with recursive logging to protect in any way the order of logging so that an unrelated thread could not insert itself between the origin of the recursion (the first LOG()) and the recursive log?

Nope, IIRC the goal was to handle the logger plugin initialization emitting its own status logs. NB: I am working from memory of the codebase from at least two years ago.

> I've also noticed some incosistencies in the relayStatusLogs function definition and its usages.

I don't think this function has "aged well" and is ready for a refactor. Sorry for the confusion, I am happy to spend time making this code path more clear and detangling the use cases.

**Comment by Smjert** [comment]
> Nope, IIRC the goal was to handle the logger plugin initialization emitting its own status logs. NB: I am working from memory of the codebase from at least two years ago.

I see, so I think I might have a fix, where each future is thread_local (and there's only "one"), so each thread waits on its own future and not the one of another thread (which was the reason for the test failing and also deadlocks happening).
Just need to test it better on Windows; I saw these lines:
https://github.com/osquery/osquery/blob/9a4e8955dc887bd2ab85828379c4aa8e3f732252/osquery/logger/logger.cpp#L353-L354
and this might be a problem if it's still true; I hope that newer MSVC versions have changed behavior.

> I don't think this function has "aged well" and is ready for a refactor. Sorry for the confusion, I am happy to spend time making this code path more clear and detangling the use cases.

Ah don't worry, I mean the async vs sync can be fixed; I was going to use an enum class with those two modes to pass to the function, so that's more explicit which is which.
That been said, if later we think there are improvement to be made, totally, I just think that might be better having it work as it was supposed to.
